src/calling/haplotypes.rs

```diff
@@ -1,48 +1,85 @@
 use anyhow::Result;
-use bio::stats::probs::LogProb;
+use bio::stats::probs::{LogProb, Prob};
 use derive_builder::Builder;
 use hdf5;
 //use kernel_density;
 //use ordered_float::OrderedFloat;
 use rust_htslib::bcf;
-//use serde_json::json;
-//use std::fs::File;
-//use std::path::PathBuf;
 use std::collections::HashMap;
+use std::io;
+use std::path::PathBuf;
 
 use bio::stats::bayesian::model::Model;
 
 use crate::haplotypes::model::{Data, HaplotypeFractions, Likelihood, Marginal, Posterior, Prior};
 
 #[derive(Builder)]
 #[builder(pattern = "owned")]
-pub(crate) struct Caller {
+pub struct Caller {
     hdf5_reader: hdf5::File,
     vcf_reader: bcf::Reader,
     min_norm_counts: f64,
+    outcsv: Option<PathBuf>,
 }
 
 impl Caller {
-    pub(crate) fn call(&self) -> Result<()> {
+    pub fn call(&self) -> Result<()> {
         // Step 1: obtain kallisto estimates.
         let kallisto_estimates = KallistoEstimates::new(&self.hdf5_reader, self.min_norm_counts)?;
 
         // Step 2: setup model.
         let model = Model::new(Likelihood::new(), Prior::new(), Posterior::new());
 
         //let universe = HaplotypeFractions::likely(&kallisto_estimates);
-        //let data = Data::new(kallisto_estimates.values().cloned().collect());
         let data = Data::new(kallisto_estimates.values().cloned().collect());
-        
+
         // Step 3: calculate posteriors.
         //let m = model.compute(universe, &data);
-        let m = model.compute_from_marginal(&Marginal::new(), &data);
+        let m = model.compute_from_marginal(&Marginal::new(3), &data);
+
+        let mut posterior = m.event_posteriors();
 
         // Step 4: print TSV table with results
         // TODO use csv crate
         // Columns: posterior_prob, haplotype_a, haplotype_b, haplotype_c, ...
         // with each column after the first showing the fraction of the respective haplotype
-
+        let mut wtr = csv::Writer::from_path(self.outcsv.as_ref().unwrap())?;
+        wtr.write_record(&[
+            "posterior_prob(log)",
+            "odds ratio",
+            "haplotype_a",
+            "haplotype_b",
+            "haplotype_c",
+        ])?; //depends upon the number of haplotypes
+
+        //write best record on top
+        let mut first_rec = Vec::new();
+        let best = posterior.next().unwrap();
+        let best_density = best.1 .0.exp();
+        let best_odds = 1;
+        first_rec.push(best_density.to_string());
+        first_rec.push(best_odds.to_string());
+        for j in 0..3 {
+            //the number depends upon the number of haplotypes
+            let fractions = best.0;
+            first_rec.push(fractions[j].to_string());
+        }
+        wtr.write_record(first_rec)?;
+
+        //write the rest of the records
+        for i in posterior {
+            let mut other_rec = Vec::new();
+            let density = i.1 .0.exp();
```

Comment:
I'd rather use "unpacking" instead of index access with tuples. E.g. in the loop above you can have:
```
for (haplotype_frequencies, prob) in posterior {
```